### ISSUE TITLE
Write MIDI from music21 `Stream` directly into `BytesIO`?

### ISSUE THREAD

**Issue by nnethery** [cuthbertLab/music21#1585]
**Motivation**

I am making a data engineering pipeline in combination of two packages, `music21` and `mido`, and I'd prefer to not have to write a midi file to the filesystem just to be imported again into `mido`. I can read/write midi data from Python STL's `io.BytesIO` object with `mido`, so it would be nice if I can write a `music21` `Stream` into a bytes buffer and construct a midi object in `mido` from that.

Trying the above gives me this error:
```
TypeError: expected str, bytes or os.PathLike object, not _io.BytesIO
```

**Feature summary**
The user can pass a `BytesIO` object from Python's native `io` library to write MIDI data into memory. 

Optionally, `converter.parse` in `music21` can read from a `BytesIO` object that is MIDI data.

**Proposed implementation**
This could probably be done inside the method `music21.midi.translate.streamToMidiFile()`, but it would be awesome to just have an interface with the standard `write()` method like this:
``` py
import io

# Assume a `music21` Score object has been created
buf = io.BytesIO()
score.write('midi', buf)
```

I see that the `streamToMidiFile()` method says it can write out to a `bytes()` object but I believe I have to construct it with the size of my midi byte representation in mind so I'm not sure how to proceed.

**Intent**

[x] I plan on implementing this myself.
[ ] I am willing to pay to have this feature added.
[ ] I am starting a discussion with the hope that community members will volunteer their time to create this. I understand that individuals work on features of interest to them and that this feature may never be implemented.

**Comment by mscuthbert** [comment]
Agreed that it would be good if .write could consistently take a Buffer object -- BytesIO for things that generate binary files and StringIO for things that generate strings -- somewhere there is already a flag on each subconverter object that records whether it generates binary output or not.

As much of the work that happens in Converter and not SubconverterMIDI or MidiStream will help other formats as well.  I don't like it when X works great in format Y but not format Z.

(related next step would be for .parse to also accept Buffers)

**Comment by nnethery** [comment]
I found that music21’s `MidiFile` class has a method called `openFileLike` that is exactly for this purpose. I can write a PR in a bit that will make the MIDI sub converter aware of it. And I’ll make sure parsing works too.

Here is a current workaround for anyone else interested in doing the same thing:
``` py
from music21.midi import translate as midiTranslate
import io

# Assume a `music21` Score object has been created

buf = io.BytesIO()
mf = midiTranslate.music21ObjectToMidiFile(score)
mf.openFileLike(buf)
mf.write()
mf.close()
```

**Comment by nnethery** [comment]
I'll work on a PR soon that will address this across all SubConverters and follow your implementation guidelines.
